### ISSUE TITLE
Names for Topos objects

### ISSUE THREAD

**Issue by craigim** [skyfielders/python-skyfield#447]
I'm writing some code that will track several `Topos` objects. It would be very helpful if I could assign a name to a location, just as I can to a `Star` and an `EarthSatellite`. Then I can query a list of `Topos` downstream to pick out a particular location by name rather than by coordinate.

**Comment by brandon-rhodes** [comment]
That’s an interesting idea! I just released a new version of Skyfield a couple of minutes ago that makes it very nearly possible, by always relying on an object’s `.target_name` for displaying the object.

The problem?

For the `Topos` I currently make that attribute read-only, because I didn’t even think of folks wanting to customize it.

Would you be willing to try an experiment on behalf of future versions of Skyfield? Then try this:
```
from skyfield import api
from skyfield.descriptorlib import reify
class MyTopos(api.Topos):
    @reify
    def target_name(self):
        return '(in this case, the normal name would print)'

site = MyTopos('31.7781 N', '35.236 E')
site.target_name = 'foo'
```

If this `MyTopos` variation on the normal `Topos` lets you set the name like you want, then I can update the real `Topos` to follow the `@reify` pattern with the next release!

**Comment by craigim** [comment]
The code successfully changes the site name using Skyfield v1.26.

**Comment by brandon-rhodes** [comment]
What about the new v1.27 — are you able to install that version and try it out? It's the new code in 1.27 that would break the ability to set `target_name` and make the above maneuver necessary.

**Comment by craigim** [comment]
I use conda for my virtual environment management and it isn't finding v1.27.

**Comment by craigim** [comment]
I will say, though, that your code snippet exactly captures what I'd like to do, except to maybe to reduce it to one step as

```python
site = Topos('31.7781 N', '35.236 E', target_name="foo")
```

**Comment by brandon-rhodes** [comment]
Oh, conda. Yes, it tends to run several hours or days behind when new software is released. Are you able to run this command?
```
pip install skyfield==1.27
```

**Comment by craigim** [comment]
If I do that, will that interfere with my environment that has v1.26 installed? The application I'm developing for is on an air-gapped computer and the version is frozen, so it's important to keep the environments sandboxed.

**Comment by brandon-rhodes** [comment]
Yes, it will install the new version — but you could return to 1.26 with `pip uninstall skyfield` followed by `conda install skyfield`, I believe?

**Comment by craigim** [comment]
The code snipped runs successfully with v1.27.

**Comment by craigim** [comment]
And if I just enter `site` in a Jupyter cell the result is `<MyTopos foo>`, `print(site)` results in `foo`.

**Comment by craigim** [comment]
but, `site = MyTopos('31.7781 N', '35.236 E', target_name="foo")` results in

```python
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-6fd5f666ee8b> in <module>
      6         return
      7 
----> 8 site = MyTopos('31.7781 N', '35.236 E', target_name="foo")
      9 # site.target_name = 'foo'
     10 site

TypeError: __init__() got an unexpected keyword argument 'target_name'


```

**Comment by brandon-rhodes** [comment]
> And if I just enter site in a Jupyter cell the result is `<MyTopos foo>`, `print(site)` results in `foo`.

Wonderful! That is how the Topos object is designed, to print a plain name when printed, but to include its class name if asked for its `repr()` (which is what a Jupyter cell asks for).

> `unexpected keyword argument`

Yes, in Python, you cannot in general provide attributes when instantiating a class. Instead, you are calling the class’s `__init__()` method which might have parameters with some of the same names as attributes on the class, but might have entirely different parameters instead.

I am not sure yet that I want to burden the `__init__()` argument list and internal code with a new argument for a use case that I am only imagining for the first time today. If I do add a way to set the displayed name at instantiation time, I will probably need to add it to all other vector functions for consistency, and I want a bit more experience with user-settable names before committing to it. (Though I'll note that the EarthSatellite already includes a parameter `name=`, so there's at least precedent established for one vector function doing so!)

But I will certainly go ahead and land a commit so that writing to `.target_name`, which became impossible in Skyfield 1.27, becomes possible again in the next version, without requiring a hand-written subclass! Interesting that you should ask for this feature just as I was breaking it.

**Comment by craigim** [comment]
`Star` also has a `names=` parameter, which is why I was surprised to see that `Topos` did not. It seems to me that I should be able to attach a name to anything I can point to or look up in an atlas.

For now I've put the custom Topos object into my code and it will make things much easier to track.

Thanks!

**Comment by brandon-rhodes** [comment]
Yes, `Star` does have an experiment in that direction — an underlying problem is that Skyfield in the future will probably only be deepening its support for vectors, because they're so much faster than loops in Python. So: the Topos object might someday soon let you simultaneously observe from *n* locations on the Earth's surface; the EarthSatellite object might support *n* satellites that all get computed at once; and so forth.

In which case I guess a `name` for the whole object might make sense, like `name='space stations'` or `name='radio observatories'`?

But the objects that have a singular name even for a collection will then have a different behavior from the oddball `Star` experiment in separate names for each star — which was an awkward move in any case because it expects a Python sequence when really stars have matured into using Pandas dataframes instead or raw Python objects and lists.

In any case: the direction I'm likely to go would be a simple `name=` for the classes that currently lack one. But for the moment I'm going to hold back until I've had a bit more experience with the new naming routines; I've often regretted it when I've innovated too quickly on the API before!

**Comment by craigim** [comment]
For context, let me give you my use case.

I've got a list of ground stations and a constellation of satellites. I use the event finder to get the rise, apex, and fall time of each satellite from each ground station. Downstream I want to separate out all contacts for station A regardless of satellite, or all contacts for satellite B regardless of station (or the union of the two lists). I can easily do the latter using the satellite name, but the former requires carrying through a separate variable for mapping location names (which I do now by sticking the Topos objects into a list of dicts) or else to have a reverse look-up table to match coordinates back to a station name.

For my purposes, looping is working just fine, though, so the current API works for me. As someone weaning themselves off of MATLAB, though, I always favor vectorization over loops.

**Comment by brandon-rhodes** [comment]
Hopefully the new `target_name` provides you with a useful alternative to the approach you have been using! I am not quite sure how to visualize the data structure that you call a "list of dicts" — that makes me think of something like:
```
[{'key': value, 'key': value, ...}, ...]
```
— but I am not sure what the keys and values would be, or why they are in a list. If you are interested in elaborating on the data structure, I might be able to suggest an alternative.

**Comment by craigim** [comment]
That is exactly what I'm doing. So I end up having a list that looks like

```python

location_list = [
    {
        "location": Topos(lat1, lon1),
        "location_id": "Kingman",
    },
    {
        "location": Topos(lat2, lon2),
        "location_id": "Barstow",
    },
    {
        "location": Topos(lat3, lon3),
        "location_id": "San Bernardino",
    },
...
]
```

Then I can use list comprehension to loop through the locations to calculate events to get

```python
contact_list = [
    {
        "location": Topos(lat1, lon1),
        "location_id": "Kingman",
        "vehicle": EarthSatellite(25544),
        "AOS_time_posix": 123456789,
        "apex_time_posix": 123456789,
        "LOS_time_posix": 123456789,
    },
    {
        "location": Topos(lat1, lon1),
        "location_id": "Kingman",
        "vehicle": EarthSatellite(25544),
        "AOS_time_posix": 124456789,
        "apex_time_posix": 124456789,
        "LOS_time_posix": 124456789,
    },
...
]
```
and then, once accumulated, turn it into a pandas table where I can sort and filter. 

```python
contact_tbl = pd.DataFrame(contact_list)
print(tbl.where(tbl.location_id == "Kingman))
print(tbl.where(tbl.vehicle.model.satnum == 25544) 
```

EarthSatellite objects and stars already carry their names with them. For Topos objects, I need to carry the extra key/val pair.

**Comment by brandon-rhodes** [comment]
Thanks for the explanation. I would perhaps have used data classes (if I was on a recent enough Python) instead of dictionaries for the same task, but the result would have been similar.

Note that Python objects allow you to assign arbitrary attributes. So even before `target_name` was writable, you could have set a `craigim_name` attribute to remember the name if you'd wanted, and avoided having a dict of two items. Instead, I think, your list could have just been the Topos objects?

In any case, I'm hoping to make a new Skyfield release within the next few days, at which point we'll be able to close this issue!

**Comment by brandon-rhodes** [comment]
As I'll be away from the keyboard next week, I think I'll avoid a Skyfield release this week lest I introduce a problem that I'm not available to correct. I'll update this issue when I return and make an official release!

**Comment by brandon-rhodes** [comment]
I have just released Skyfield 1.28 — enjoy!
